From 595c732c1c2ae3635dc2faf5757a61d7a5a5e44d Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Tue, 31 Jan 2023 19:34:47 +0100
Subject: [PATCH 272/841] fix(st-ddr): stm32mp2: ensure correct LPDDR4
 self-refresh configuration

Enable PWRCTL.SELFREF_SW before DDRCTRL static configuration to ensure
correct setting of PWRCTL.LPDDR4_SR_ALLOWED.
Later disabled in restore_refresh() function at the end of the
initialization sequence.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git b0d4940cbc55b6778c4488fbc26742f70fa407f9 fix(st-ddr): stm32mp2: ensure correct LPDDR4 self-refresh configuration]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: Id24607c32b2ae4babfacc7de4acc5ec91925a1b4
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/288317
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/ddr/stm32mp2_ddr.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/st/ddr/stm32mp2_ddr.c b/drivers/st/ddr/stm32mp2_ddr.c
index 7d386a971..9a1c3c259 100644
--- a/drivers/st/ddr/stm32mp2_ddr.c
+++ b/drivers/st/ddr/stm32mp2_ddr.c
@@ -338,6 +338,12 @@ static void restore_refresh(struct stm32mp_ddrctl *ctl, uint32_t rfshctl3, uint3
 		udelay(DDR_DELAY_1US);
 	}
 
+	if ((pwrctl & DDRCTRL_PWRCTL_SELFREF_SW) != 0U) {
+		mmio_clrbits_32((uintptr_t)&ctl->pwrctl, DDRCTRL_PWRCTL_SELFREF_SW);
+
+		udelay(DDR_DELAY_1US);
+	}
+
 	if ((pwrctl & DDRCTRL_PWRCTL_POWERDOWN_EN) != 0U) {
 		mmio_setbits_32((uintptr_t)&ctl->pwrctl, DDRCTRL_PWRCTL_POWERDOWN_EN);
 
@@ -410,6 +416,14 @@ void stm32mp2_ddr_init(struct stm32mp_ddr_priv *priv,
 		ddr_sysconf_configuration(priv);
 	}
 
+#if STM32MP_LPDDR4_TYPE
+	/*
+	 * Enable PWRCTL.SELFREF_SW to ensure correct setting of PWRCTL.LPDDR4_SR_ALLOWED.
+	 * Later disabled in restore_refresh().
+	*/
+	config->c_reg.pwrctl |= DDRCTRL_PWRCTL_SELFREF_SW;
+#endif /* STM32MP_LPDDR4_TYPE */
+
 	stm32mp_ddr_set_reg(priv, REG_REG, &config->c_reg, ddr_registers);
 	stm32mp_ddr_set_reg(priv, REG_TIMING, &config->c_timing, ddr_registers);
 	stm32mp_ddr_set_reg(priv, REG_MAP, &config->c_map, ddr_registers);
-- 
2.47.2

